drivers/st25r3911b/nfca.c

```diff
@@ -0,0 +1,1236 @@
+	      ST25R3911B_IRQ_MASK_FWL;
+
+	if (fdt != 0) {
+		irq |= ST25R3911B_IRQ_MASK_NRE | ST25R3911B_IRQ_MASK_GPE;
+
+		err = transfer_fdt_set(fdt);
+		if (err) {
+			return err;
+		}
+	}
+
+	err = transmission_prepare();
+	if (err) {
+		return err;
+	}
+
+
+	/* Enable TX and RX interrupts */
+	err = st25r3911b_irq_enable(irq);
+	if (err) {
+		return err;
+	}
+
+	/* Write Tx data to FIFO, if data is longer than FIFO write,
+	 * then only write first 96 bytes.
+	 */
+	err = st25r3911b_fifo_write(tx->data,
+				    MIN(tx->len, ST25R3911B_MAX_FIFO_LEN));
+	if (err) {
+		return err;
+	}
+
+	/* Set Rx buffer */
+	nfca.transfer.rx_buf = rx;
+
+	/* Set data length to send */
+	err = st25r3911b_tx_len_set(tx->len);
+	if (err) {
+		return err;
+	}
+
+	if (auto_crc) {
+		cmd = ST25R3911B_CMD_TX_WITH_CRC;
+
+		LOG_DBG("Transmit data with auto generated CRC");
+
+	} else {
+		cmd = ST25R3911B_CMD_TX_WITHOUT_CRC;
+
+		LOG_DBG("Transmit data without CRC");
+	}
+
+	nfca.transfer.written_byte = MIN(tx->len, ST25R3911B_MAX_FIFO_LEN);
+
+	return st25r3911b_cmd_execute(cmd);
+}
+
+static int tag_sleep(void)
+{
+	int err;
+	u8_t tx_buffer[] = NFCA_SLEEP_CMD;
+	struct nfca_buf tx = {
+		.data = tx_buffer,
+		.len = sizeof(tx_buffer)
+	};
+
+	err = transfer(&tx, NULL, 0, true);
+	if (err) {
+		return err;
+	}
+
+	nfca.tag.sleep = true;
+
+	return 0;
+}
+
+static int user_state_process(void)
+{
+	int err = 0;
+
+	switch (atomic_get(&nfca.state.tag)) {
+	case STATE_FIELD_ON:
+		err = st25r3911b_field_on(ST25R3911B_NO_THRESHOLD_ANTICOLLISION,
+					  ST25R3911B_NO_THRESHOLD_ANTICOLLISION,
+					  NFCA_GT_TIME);
+
+		state_set(STATE_IDLE);
+
+		if (!err) {
+			if (nfca.cb->field_on) {
+				nfca.cb->field_on();
+			}
+		}
+
+		break;
+
+	case STATE_FIELD_OFF:
+		err = st25r3911b_rx_tx_disable();
+
+		state_set(STATE_IDLE);
+
+		if (!err) {
+			if (nfca.cb->field_off) {
+				nfca.cb->field_off();
+			}
+		}
+
+		break;
+
+	case STATE_TAG_DETECTION:
+		err = tag_detect(nfca.cmd);
+		break;
+
+	case STATE_ANTICOLLISION:
+		err = anticollision_resolution(ANTICOLLISION_CASCADE_1);
+		break;
+
+	case STATE_TRANSFER:
+		err = transfer(nfca.transfer.tx_buf,
+			       nfca.transfer.rx_buf,
+			       nfca.transfer.fdt,
+			       nfca.transfer.auto_crc);
+		break;
+
+	case STATE_SLEEP:
+		err = tag_sleep();
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+static void timeout_handler(struct k_work *work)
+{
+	timeout_process();
+}
+
+void st25r3911b_nfca_process(void)
+{
+	int err = 0;
+
+	if (nfca_events[NFCA_USER_EVENT_IDX].state == K_POLL_STATE_SEM_AVAILABLE) {
+		k_sem_take(nfca_events[NFCA_USER_EVENT_IDX].sem, K_NO_WAIT);
+
+		LOG_DBG("User state process");
+
+		err = user_state_process();
+
+		nfca_events[NFCA_USER_EVENT_IDX].state =
+					K_POLL_STATE_NOT_READY;
+	}
+
+	if (nfca_events[NFCA_IRQ_EVENT_IDX].state == K_POLL_STATE_SEM_AVAILABLE) {
+		k_sem_take(nfca_events[NFCA_IRQ_EVENT_IDX].sem, K_NO_WAIT);
+
+		LOG_DBG("Interrupt process");
+
+		err = irq_process();
```

Comment:
Remember that you can have to handle both events.
In such a situation if `user_state_process` returns an error but `irq_process` will return zero, error  will be ignored.
Also I think @lemrey  is right and in top level API you should return an error instead of assert.

Reply:
Ok right. Fixed ;)